### DISCUSSION TITLE
Enhancing support for audiobooks

### DISCUSSION THREAD

**Discussion by Neurrone** [beetbox/beets#4269] [Ideas]
Hi,

I would love to be able to use Beets to manage my audiobook collection. I'm writing a plugin to fetch book data from Audible and find that it works very well when the audiobook is a folder of mp3s, where each mp3 is a chapter of the book.

There are two main scenarios that beets doesn't yet support, and I'm wondering if there are workarounds for this or if this is a feature that could be added?

1. M4b files with embedded chapter markers: this is probablyy the most common case, storing it in a single file is convenient for consumption and for file copying speeds.
    
    Can Beets somehow recognize that the file actually contains tracks?
2. Audio that is not split by chapter. This could be a single mp3, a single m4b without embedded chapter markers or a folder of many mp3s. Or this could also be a folder with subfolders (e.g, disk 1, disk 2 etc).
    
    If possible, I would want Beets to somehow not care about it not being organized by chapter but still apply the correct metadata to everything in that book. Or have some way of managing files that don't have a one-one relationship with tracks.

**Comment by sampsyo** [comment]
Both of these seem super interesting, and it would be great to support them! Neither one, however, is really supported in beets currently—neither mapping multiple "tracks" into a single file nor mapping a single "track" to multiple files.

They remind me a bit of #136, which is really old, and for which demand has really fallen off in recent years—but the idea there also is that people want to rip a CD as a single, continuous audio file and then logically view it as multiple tracks in beets.

Anyway, these both seem like interesting but ambitious projects. They would require some significant internal reorganization. So it might be best to start by getting familiar with the data structures and database layer in beets to understand what the scope would be.

**Comment by Neurrone** [comment]
It feels like supporting the first case, single file being able to represent an entire album would be easier to implement / start with.

Where should I start to look and at a high level, what would need to be done to support this?

**Reply by sampsyo** [reply]
It's pretty hard to say exactly what the strategy would be here that could succeed. However, I think the main place to start looking & thinking would be at the database model. Namely, you might want to start with the `Item` class in the `library` module, which is a single row in the database representing a single track—and, currently, has a 1:1 association with a file on disk. Trying to think through what it would look like to change this correspondence to N:1 would be a good place to start.

**Reply by Neurrone** [reply]
Focusing on having a file being able to represent multiple tracks, I see that the `items` table has a `path` attribute. Here are some preliminary ideas I have on how to proceed:

1. Each item gets two new columns, `start_duration` and `end_duration`.
	* We could have these columns be nullable in which case, nulls in both columns mean that the track spans the entire file.
	* Alternatively, these could be not nullable, and so having `start_duration=0` and `end_duration=length` would represent the same thing. I prefer this approach. The length column would technically be redundant and could be removed if desired.
2. Update functions that read a track from disk.
	* If `start_duration == 0` and `end_duration` matches the duration of the file read from the disk, then the track spans the whole file. Fall back to existing logic.
		* I'm assuming that the duration of the file can be read efficiently, e.g that its stored near the start of the file
	* Otherwise, fetch the relevant segment of data from the file and return it
3. Extend the importer to recognize chapter markers in file formats which support them such as m4b and mp3 so the single file gets interpreted as multiple tracks
4. Writing id3 tags would only need to be done once per file, not once per track.
	* If a file represents say 5 tracks, it will be possible for each track to have differing ID3 tags. We would need to decide which data to use to write to the enclosing file when they don't agree.
5. Deletion of a file from disk can now only happen when no tracks reference any part of it.
6. This will break backwards compatibility with events that assume an item maps to a single file:
	* `item_copied`: no longer called once per track, this will only be called once per file. Or there could be a separate event emited called `track_copied`, which would be emited each time a track in the same file gets copied? Likewise for the rest of thee events
	* `before_item_moved`
	* `item_moved`
	* `item_linked`
	* `item_hardlinked`
	* `item_reflinked`
	* `item_removed`
	* `write`
	* `after_write`

	I'm thinking it may make sense to rename these events to e.g, `track_write` and create new ones specifically for the file write.
7. Path formats will also need to be updated to be aware of the track / file distinction

Let me know if I missed anything.

As you said, this is a big change, curious on your thoughts on whether you'd like to proceed given the additional complexity this will add to the codebase?

After looking at what is needed to have a file map to n tracks, I think having a track map to n files would be even more complex and I will shelve that idea for now, as the extra utility that provides on top of this is minimal.

**Reply by sampsyo** [reply]
Wow, nice, thanks for thinking this through! Something like this does seems like the right way to proceed, and it seems cool to investigate it.

Here’s one difficult question: what would be a minimally disruptive way to start? It would be really great to break down the implementation into phases, which can roughly align with the items in your list above. But even better if we can pick some changes that will take steps toward the broader feature without changing too much of the code all at once. For example, adding those start/end fields but preserving 1:1 item:file correspondence, while enabling item 2 above, could be interesting by itself and wouldn’t require much change.

**Reply by Neurrone** [reply]
1. Phase 1:
	* Adding both new columns
	* The read track functions can be updated to add the new logic to return subsections of a file, but because the track = file correspondance is still there, it should be backwards compatible. So the main work here is figuring out how to return the requested audio data based on start / end.
2. Phase 2: in this phase, the user shouldn't notice any changes in behaviour if they are not importing files with chapter markers. We could also gate the new functionality behind a feature flag.
	* Extend the i/o layer to recognize chapter markers in mp3 and m4b files and present them as separate tracks.
	* Update the writing infrastructure to be aware of the file / track distinction.
		* What happens if the chapter markers don't match track data fetched? I'm wondering how Beets currently handle duration mismatches. In this case, we have an option to correct these chapter markers if desired
		* Decide how to write id3 data to the enclosing file
	* Update file deletion so that it only happens when all tracks that reference it are deleted
	* Emit `file_*` events and figure out what parameters should be passed into listeners
3. Phase 3: start making backwards incompatible changes:
	* Update plugin events, I think it may be better to rename `item_*` to `track_*`.
	* Update path functions

**Reply by sampsyo** [reply]
Neat! Phase 2 seems like the hard one, to be honest, and we might need to break it down even further. But I'm down to proceed with the design for Phase 1…

**Reply by Neurrone** [reply]
Great! Starting with something simpler will also help me learn the codebase.

Some ideas / questions for starting to work on phase 1.

1. How does Beets read files and audio metadata from disk? E.g, getting details like duration of the file? Am looking for the code that does this to figure out how to return segments of a file.
2. Any preferences on how the database migration should be performed? What I have in mind is the 2 new non-nullable columns for start and end duration. 
	
	* I noticed that length is a float in seconds, I was thinking of using an integer to store the start and end duration in ms instead. I usually default to integers when I can to avoid dealing with the complexity of floats.
	* Would this be a good time to store file information in a separate `files` table? Or keep `path` as is for now.
	* Would you want to keep the `length` column, or remove it because it can be deduced from `end - start`?

**Reply by sampsyo** [reply]
All the responsibilities for item 1 belong to [MediaFile](https://github.com/beetbox/mediafile), which wraps the low-level data maculation library [Mutagen](https://mutagen.readthedocs.io/en/latest/).

I think the questions about the database schema might require a somewhat longer conversation… maybe that deserves a new discussion thread of its own?

**Reply by wisp3rwind** [reply]
> * This will break backwards compatibility with events that assume an item maps to a single file:

This will actually break plugins in all kinds of ways (hooks, import stages, plugin commands, ...); the expectation of a one-to-one mapping between `Item`s and media files is probably very deeply ingrained in the beets ecosystem. I would expect that this can easily lead to data loss due to unsuspecting plugins. Thus, planning this feature should also involve a plan on how to release it. Some things to consider:

- Step major version number
- Communication with plugin authors and packagers
- Think about implementing a mechanism to detect plugins that don't explicitly support "Beets 2.0" (this would be essential in my opinion, and in general nice to have should we want to break plugins again).
- Since this will involve a huge number of changes, think about which preparatory parts of the implementation can already be merged and shipped before actually enabling 1-to-N mapping of items.
- Maybe use the opportunity to break plugin API even further and remove some legacy code and unfortunate design decisions.

This is not to say it can't be done, but it is a big project.

**Comment by Neurrone** [comment]
I've pushed a preliminary version of an [Audible plugin](https://github.com/Neurrone/beets-audible).

I hope to have more time in a few weeks to investigate extending Beets to better support this use case.

**Comment by Neurrone** [comment]
On further thought, I realize I've been making the assumption of being able to get chapter metadata (i.e, details for each track) for all books, which is probably not going to be the case. I am sure that I can get info on the overall book though.

I'm wondering if its possible to have Beets somehow support cases where track-level metadata isn't available and there's only overall info on the album? What I'd want to have is:

1. Some way of having Beets match tracks that are in the same book, even if my plugin isn't able to provide metadata at the level of individual chapters. So for example, this means the number of tracks in the album is unknown. I would have details on the book, description, authors etc though.
2. Write info about the book to ID3 tags on each track.

Is this currently possible to achieve within the existing plugin system? If so, I think this may be an easier interim step to supporting audiobooks that aren't properly chapterized. 

* If chapter-level metadata is available and if the number of tracks doesn't match, then assume the book has not been chapterized, and don't return track-level metadata
* Without track-level matching, it falls back to the scenario I described above where matching is only done based on the album.

**Reply by Neurrone** [reply]
@sampsyo any thoughts on whether this might be possible?

**Reply by sampsyo** [reply]
That's a good question. It doesn't immediately seem impossible to me… one could just imagine filling the `tracks` field of an `AlbumInfo` object with dummy objects? Maybe that's worth at least experimenting with?

**Reply by Neurrone** [reply]
I'm thinking that I would need to return a list of dummy objects (filled in based on album / book-level info) so that its length is the same as the number of files in that folder. Is there any way of getting that info? Since the events don't pass this information in.

I'm assuming there is a penalty or the match will be harder if the number of track info objects don't match the number of files on disk being matched to that album.

**Reply by sampsyo** [reply]
The `candidates` plugin hook, which plugins use to provide metadata matches, gets an `items` list as an argument.

**Reply by Neurrone** [reply]
Thanks so much, I'm not sure how I didn't realize this.

This has facilitated a massive breakthrough, I believe I can get Beets to work well with virtually all audiobooks now. So much so that I feel the complexity from having Beets support multiple tracks per file wouldn't be worth the effort, unless there are other use cases besides audiobooks.

The key insight is to tell a white lie to Beets when the audio being added isn't chapterized. For example, if the entire book is in one file, I would return a single `track` instead of however many chapters the book actually has.

I'm still polishing the code but will update here when its ready for wider consumption.

**Reply by sampsyo** [reply]
That's awesome to hear!

**Reply by Neurrone** [reply]
its ready for use, see https://github.com/Neurrone/beets-audible

Let me know if the instructions are unclear. I need to create a mod for the linuxserver Beets container to make setting it up with Docker easier.

I've tested this on a variety of books and so far, it has worked on all of them.

**Reply by sampsyo** [reply]
Awesome. Are you interested in adding a link to the docs, in our list of third-party plugins? https://github.com/beetbox/beets/blob/master/docs/plugins/index.rst#other-plugins

**Reply by Neurrone** [reply]
Yup, I'll do so either this weekend or next.

I'd like to rename the plugin to audiobooks instead of audible just in case it grows to support additional sources of metadata.

**Comment by Neurrone** [comment]
Some quick questions:

* Is there a way for me to associate additional files to the album besides the cover art? I notice that if I edit the library to change e.g, the title or author of the album, the audio and cover gets correctly moved to the new location but there are additional files I write to the directory on initial import that get left behind.
* When importing an already chapterized book, sometimes it tries to make corrections to the track metadata that would mistag a chapter to the wrong chapter. For example, trying to change chapter 5 to chapter 7 because the length appears similar. Is there a way for me to ignore this change and have it proceeding with the import, but at the same time, apply other types of metadata like album artist, comments etc?
* I notice via verbose logging that most of the time, many candidates are filtered out because Beets detects search results that are similar as being duplicates even though they're not in this case. Is there a way for me to disable this behaviour?

**Reply by sampsyo** [reply]
There is not currently a way to associate additional files. That is the longstanding "attachments" proposal (e.g., #591) that has proved more difficult than expected to implement. It is also the topic of [beets-copyartifacts](https://github.com/adammillerio/beets-copyartifacts).

I also can't currently think of a way to partially apply a match—i.e., just its album/albumartist but not any track metadata. Sounds potentially useful!

The matcher shouldn't be filtering out similar-but-not-identical matches. It just uses the ID—is there any chance these matches have the same ID?

**Reply by Neurrone** [reply]
> The matcher shouldn't be filtering out similar-but-not-identical matches. It just uses the ID—is there any chance these matches have the same ID?

Which ID are you referring to? Is it the musicbrainze album id? I've set that to `None` because that seemed to be the most reasonable value in this instance, likewise for this id for all the tracks.

**Reply by sampsyo** [reply]
Yes. Give that a shot, at least—it's a design flaw, which has also proven very difficult to resolve, that this is the ID used to deduplicate matches (cf #604).

**Reply by Neurrone** [reply]
Do you mean assigning a unique musicbrainz id? I do have the asin which is unique.

Another temporary solution I could think of is to disable duplicate detection if the id is None, as it doesn't make sense for it to be done in this situation.

**Reply by sampsyo** [reply]
Yes, exactly. Both of those would be good options.
